### ISSUE TITLE
[Regression] Conda create environment fails on lock if root environment is not under user control

### ISSUE THREAD

**Issue by frol** [conda/conda#2681]
This issue is introduced in Conda 4.1.0 (Conda 4.0.8 works fine).

```
$ conda create -n root2 python=2                                                                                                                                                                   [123/1811]
Fetching package metadata .......
Solving package specifications .............

Package plan for installation in environment /home/frol/.conda/envs/root2:

The following NEW packages will be INSTALLED:

    openssl:    1.0.2h-1      (soft-link)
    pip:        8.1.2-py27_0  (soft-link)
    python:     2.7.11-0      (soft-link)
    readline:   6.2-2         (soft-link)
    setuptools: 23.0.0-py27_0 (soft-link)
    sqlite:     3.13.0-0      (soft-link)
    tk:         8.5.18-0      (soft-link)
    wheel:      0.29.0-py27_0 (soft-link)
    zlib:       1.2.8-3       (soft-link)

Proceed ([y]/n)?

Linking packages ...
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "/usr/local/miniconda/bin/conda", line 6, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/main.py", line 120, in main
    args_func(args, p)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/main.py", line 127, in args_func
    args.func(args, p)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/main_create.py", line 57, in execute
    install(args, parser, 'create')
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/install.py", line 407, in install
    execute_actions(actions, index, verbose=not args.quiet)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/plan.py", line 566, in execute_actions
    inst.execute_instructions(plan, index, verbose)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/instructions.py", line 137, in execute_instructions
    cmd(state, arg)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/instructions.py", line 80, in LINK_CMD
    link(state['prefix'], dist, lt, index=state['index'], shortcuts=shortcuts)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/install.py", line 1035, in link
    with Locked(prefix), Locked(pkgs_dir):
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/lock.py", line 60, in __enter__
    os.makedirs(self.lock_path)
  File "/usr/local/miniconda/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/usr/local/miniconda/pkgs/.conda_lock-949'
```

`/usr/local/miniconda/` is a system-wide installation of miniconda, so obviously, users cannot create lock files there.

P.S. I have a dream that updating conda software won't break things on every release...

**Comment by frol** [comment]
It seems that I cannot even do `source activate ...` as a regular user now. It just hangs. Here is what I get when I interrupt it with `^C`:

```
$ source activate root2
^CTraceback (most recent call last):
  File "/usr/local/miniconda/bin/conda", line 6, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/main.py", line 48, in main
    activate.main()
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/activate.py", line 121, in main
    path = get_path(shelldict)
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/cli/activate.py", line 108, in get_path
    return run_in(shelldict["printpath"], shelldict)[0]
  File "/usr/local/miniconda/lib/python2.7/site-packages/conda/utils.py", line 174, in run_in
    stdout, stderr = p.communicate()
  File "/usr/local/miniconda/lib/python2.7/subprocess.py", line 799, in communicate
    return self._communicate(input)
  File "/usr/local/miniconda/lib/python2.7/subprocess.py", line 1409, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/local/miniconda/lib/python2.7/subprocess.py", line 1463, in _communicate_with_poll
    ready = poller.poll()
KeyboardInterrupt
```

**Comment by frol** [comment]
It also seems like there is no way to pin Conda version. Installing any package to the root env tries to update Conda to the latest version:

```
$ conda install python
Using Anaconda Cloud api site https://api.anaconda.org
Fetching package metadata: ....
Solving package specifications: .........

Package plan for installation in environment /usr/local/miniconda:

The following packages will be UPDATED:

    conda:     4.0.8-py27_0 --> 4.1.0-py27_0
    conda-env: 2.4.5-py27_0 --> 2.5.0-py27_0

Proceed ([y]/n)?
```

**Comment by kalefranz** [comment]
You can now pin 4.0.9.

    conda install conda=4.0.9
    conda config --set auto_update_conda false

`auto_update_conda` setting will be added to 4.1.1 also, which is coming out tonight or tomorrow morning.

**Comment by kalefranz** [comment]
Freaking lock on the whole package cache.  We're going to get rid of that soon.  I can't really tell what initially tripped here.  What are the permissions on `/usr/local/miniconda/pkgs/.conda_lock-949`?  Are they yours, or root, or does it even exist?

**Comment by kalefranz** [comment]
Ohhh I think I get it, from your issue title.  I actually thought this was already broken anyway.  I plan on having 4.2 out in a couple weeks now, and should be a simple fix at that point.  Ok with staying on 4.0.9 for this use case until then?

**Comment by frol** [comment]
@kalefranz It seems that I don't have options here but wait. I use Docker containers and this bug doesn't bother me that much. However, I would say that it is not a minor regression to postpone it to the next release. (From my experience, next release will break things in another way, so people will stuck with 4.0.*)

**Comment by kalefranz** [comment]
In the last several months our coverage has gone from ~48% to almost 70%. The code base is still far more fragile and brittle than I'd like it to be, but we're making progress I think.

**Comment by frol** [comment]
Reverting #2320 fixed the regression. However, it seems that it just fails silently at locking, but at least it works in non-concurrent scenarious.

cc @alanhdu

**Comment by alanhdu** [comment]
@frol: Yeah, that's about right. Before #2320, conda would just swallow all `OsError`s (including `PermissionError`s). 

For now, we could add a `try ... except` around https://github.com/conda/conda/blob/master/conda/lock.py#L60, catch a `PermissionError` (or whatever the equivalent Python 2 error is), and try to do something smart (or just silently fail... depends how important that lock actually is).

**Comment by damianavila** [comment]
I am seeing this issue as well @kalefranz

**Comment by davidslac** [comment]
After reading this, I thought I would be able to downgrade conda to 4.0.9. However after doing that, I am still unable to activate centrally administered conda environments as a user. Is there a prior version of conda you would reccommend? Or did the 4.1.3 version leave something in my miniconda install that is causing the problem? Do I need to re-install miniconda2 from scratch? Should I just wait for this to be fixed before proceeding with trying to build a central conda install for our users?

**Comment by frol** [comment]
@davidslac There are two options you may try:

1. Downgrade conda-env:

  ```bash
  $ conda install 'conda-env<2.5'
  ```
2. Patch (revert changes) conda lock:

  ```bash
  $ curl -o "/usr/local/miniconda/lib/python"*"/site-packages/conda/lock.py" \
           "https://raw.githubusercontent.com/conda/conda/9428ad0b76be55e8070e04dd577c96e7dab571e0/conda/lock.py"
  ```

**Comment by davidslac** [comment]
Thanks! I tried both, but it sill did not work. It's probably something 
I'm overlooking on my part - after the first failure I deleted the 
lock.pyc, but still no luck.

I'll just hope for a fix in the future. I suspect though, that central 
administration is not as standard of a use case, so one is more likely 
to run into problems. It may be we should just provide a channel of our 
packages to our users and let them administer their own software stacks, 
give them some environments we know work.

best,

David
